lib/compress/huf_compress.c

```diff
@@ -677,28 +705,56 @@ static void HUF_buildCTableFromTree(HUF_CElt* CTable, nodeElt const* huffNode, i
     CTable[0] = maxNbBits;
 }
 
-size_t HUF_buildCTable_wksp (HUF_CElt* CTable, const unsigned* count, U32 maxSymbolValue, U32 maxNbBits, void* workSpace, size_t wkspSize)
+static size_t
+HUF_nbSymbolsTooLarge(const nodeElt* hnodes, U32 maxSymbolValue, U32 maxNbBits)
+{
+    size_t nbSTL = 0;
+    int s = (int)maxSymbolValue;
+    for ( ; s > 0; s-- ) {
+        if (hnodes[s].nbBits > maxNbBits) nbSTL++;
+        else break;
+    }
+    return nbSTL;
+}
+
+size_t
+HUF_buildCTable_wksp(HUF_CElt* CTable, const unsigned* count, U32 maxSymbolValue, U32 maxNbBits,
+                     void* workSpace, size_t wkspSize)
 {
-    HUF_buildCTable_wksp_tables* const wksp_tables = (HUF_buildCTable_wksp_tables*)HUF_alignUpWorkspace(workSpace, &wkspSize, ZSTD_ALIGNOF(U32));
+    HUF_buildCTable_wksp_tables* const wksp_tables =
+        (HUF_buildCTable_wksp_tables*)HUF_alignUpWorkspace(workSpace, &wkspSize, ZSTD_ALIGNOF(U32));
     nodeElt* const huffNode0 = wksp_tables->huffNodeTbl;
     nodeElt* const huffNode = huffNode0+1;
     int nonNullRank;
 
+    DEBUGLOG(5, "HUF_buildCTable_wksp (alphabet size = %u)", maxSymbolValue+1);
+
     /* safety checks */
     if (wkspSize < sizeof(HUF_buildCTable_wksp_tables))
-      return ERROR(workSpace_tooSmall);
+        return ERROR(workSpace_tooSmall);
     if (maxNbBits == 0) maxNbBits = HUF_TABLELOG_DEFAULT;
     if (maxSymbolValue > HUF_SYMBOLVALUE_MAX)
-      return ERROR(maxSymbolValue_tooLarge);
+        return ERROR(maxSymbolValue_tooLarge);
     ZSTD_memset(huffNode0, 0, sizeof(huffNodeTable));
 
     /* sort, decreasing order */
     HUF_sort(huffNode, count, maxSymbolValue, wksp_tables->rankPosition);
+    DEBUGLOG(6, "sorted symbols completed (%zu symbols)", showHNodeSymbols(huffNode, maxSymbolValue+1)); (void)showHNodeSymbols;
 
     /* build tree */
     nonNullRank = HUF_buildTree(huffNode, maxSymbolValue);
 
-    /* enforce maxTableLog */
+    /* determine and enforce maxTableLog */
+    /* Loosen target when maxNbBits is already within limits.
+     * A harsh rebalancing can be bad for compression ratio
+     * while a mild one tends to be better */
+    while (maxNbBits < HUF_TABLELOG_DEFAULT) {
+        size_t const nbSTL = HUF_nbSymbolsTooLarge(huffNode, maxSymbolValue, maxNbBits);
+        #define HUF_NB_NODES_TO_FIX_MAX 32
+        if (nbSTL < HUF_NB_NODES_TO_FIX_MAX) /* heuristic */
+            break;
+        maxNbBits++;
+    }
```

Comment:
This is breaking the fuzzer, because it expects that if it passes the `maxNbBits` to `HUF_buildCTable_wksp()`, it won't build a table with a larger table log. And it picks a random tableLog between the minimum possible and 12.

I guess the table log gets shrunk too far by `HUF_optimalTableLog()`. Would it make sense to move the call to `HUF_optimalTableLog()` into this function? Or replace `HUF_optimalTableLog()` with the logic here, and shrink where applicable, instead of growing?

I think logically, we should only be modifying the table log in one place anyway. So merging this wit h`HUF_optimalTableLog()` makes sense.

Reply:
Thought about that too.
The issue is that `HUF_optimalTableLog()` doesn't have enough information to make such choice. `HUF_optimalTableLog()` only receives `srcSize` and `maxSymbolValue` to make a decision.
In this PR, the proposed refinement can happen thanks to access to `huffNode`, which is a sorted list of symbols with already decided `nbBits` (no maximum enforced yet). This information is not available anywhere before, so it can't be used before that point, and `HUF_optimalTableLog()` is run before even entering this function.

There are multiple options starting from there, with more or less refactoring cost.

One that comes to mind is that the test is simply no longer appropriate.
What it tests is that, if the huffman compressor is instructed to use a low (yet achievable) `tableLog`, then it will always respect the limit, no matter the cost to compression ratio.
However, if we accept the premisses that the huffman codec within `zstd` is no longer "generic" but designed solely for `zstd`, then some of its existing properties are no longer relevant. In this instance, `HUF_buildCTable_wksp()` does no longer guarantee to respect the maximum `tableLog` provided, though it does guarantee to remain below `tableLog==11`, which is a limit specific to `zstd` proper. The test could simply adjust to the new contract.

One issue I see with this design is that it's no longer possible to "force" `HUF_buildCTable_wksp()` to use an explicit low `tableLog`, because the function is free to decide on its own that it's too low, and fix it to use its own limit. Yet, even the new heuristic, while better informed than `HUF_optimalTableLog()`, is still just an heuristic, and could be wrong. A more thorough analysis might actually discover that it's indeed better to use a smaller table. Or it might want to use a smaller table for other reasons while accepting a compression ratio cost (speed of table construction for example).